#615 - listener for assignments

nachogiljaldo opened the pull request:
Closes https://github.com/segmentio/kafka-go/issues/615

Adds a listener to send re-assignments to a listener.

nachogiljaldo commented:
@achille-roussel I saw you commented on https://github.com/segmentio/kafka-go/issues/615 . Is this something you could review?

isaacd9 commented:
I don't think adding this is a good idea. Instead, you should probably use the [ConsumerGroup](https://pkg.go.dev/github.com/segmentio/kafka-go#ConsumerGroup) API to listen for a new generation and construct a new reader.

nachogiljaldo commented:
Thanks for the feedback @isaacd9 ! Could you elaborate?

I had a quick look at the [NewReader](https://github.com/segmentio/kafka-go/blob/main/reader.go#L629). Also it seems to use `run` which is private, so it's not possible to do it that way without duplicating even more code? Same for `getTopics()` while building the ConsumerGroupConfig.

Maybe, instead, a modification safe version of ConsumerGroup should be exposed (that only has Next but not Close) in the `Reader`? That allows to preserve the encapsulation of the logic to create Readers which IMHO seems important and avoids unnecessary duplication on all clients that would need this.

Something like:
```
type ConsumerGroupGenerationListener interface {
  Next(ctx context.Context) (*Generation, error)
}

...

func (r *Reader) ConsumerGroupGenerationListener(ctx context.Context) (ConsumerGroupGenerationListener, error) {
	if !r.useConsumerGroup() {
		return nil, errOnlyAvailableWithGroup
	}
        return r.consumerGroupGenerationListener, nil
}
```

nachogiljaldo commented:
@petedannemann , sorry for the direct ping but you seem to have reviewed some of the last PRs. I wonder if this is something you could provide feedback on the approach (or the alternatives provided by Isaac or my counter proposal).

nachogiljaldo commented:
heya @erikdw / @jkoelker apologies for the direct ping, but this has been laying around for a while and I would need some guidance (as you seem to have approved the last merged PRs)

If you're indeed maintainers of this library, are you ok with this approach? Would you prefer the one outlined in https://github.com/segmentio/kafka-go/pull/1305#issuecomment-2205206085 or a different one? Happy to take on those suggestions

nachogiljaldo commented:
@jkoelker thanks for the prompt reply!

Fair points, I updated the PR description and first commit message to be more informative, I hope that makes sense.

Also removed the sorting as you suggested.

As per the specifics of the implementation, since you said you're not familiar with the code or kafka, would you know who could help providing feedback on the approach / review this PR?
